src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/JNativeScanTask.java

```diff
@@ -0,0 +1,193 @@
+        this.out = out;
+        this.classPaths = classPaths;
+        this.modulePaths = modulePaths;
+        this.version = version;
+        this.action = action;
+        this.cmdRootModules = cmdRootModules;
+    }
+
+    public void run() throws JNativeScanFatalError {
+        List<ScannedModule> modulesToScan = new ArrayList<>();
+        findAllClassPathJars().forEach(modulesToScan::add);
+
+        ModuleFinder moduleFinder = ModuleFinder.of(modulePaths.toArray(Path[]::new));
+        List<String> rootModules = cmdRootModules;
+        if (rootModules.contains("ALL-MODULE-PATH")) {
+            rootModules = allModuleNames(moduleFinder);
+        }
+        Configuration config = Configuration.resolveAndBind(moduleFinder, List.of(systemConfiguration()),
+                ModuleFinder.of(), rootModules);
+        for (ResolvedModule m : config.modules()) {
+            URI location = m.reference().location().orElseThrow();
+            Path path = Path.of(location);
+            checkRegularJar(path);
+            modulesToScan.add(new ScannedModule(path, m.name()));
+        }
+
+        Map<ScannedModule, Map<ClassDesc, List<RestrictedUse>>> allRestrictedMethods;
+        try(ClassResolver classesToScan = ClassResolver.forScannedModules(modulesToScan, version);
+                ClassResolver systemClassResolver = ClassResolver.forSystemModules(version)) {
+            NativeMethodFinder finder = NativeMethodFinder.create(classesToScan, systemClassResolver);
+            allRestrictedMethods = finder.findAll();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        switch (action) {
+            case PRINT -> printNativeAccess(allRestrictedMethods);
+            case DUMP_ALL -> dumpAll(allRestrictedMethods);
+        }
+    }
+
+    // recursively look for all class path jars, starting at the root jars
+    // in this.classPaths, and recursively following all Class-Path manifest
+    // attributes
+    private Stream<ScannedModule> findAllClassPathJars() throws JNativeScanFatalError {
+        Stream.Builder<ScannedModule> builder = Stream.builder();
+        Deque<Path> classPathJars = new ArrayDeque<>(classPaths);
+        while (!classPathJars.isEmpty()) {
+            Path jar = classPathJars.poll();
+            checkRegularJar(jar);
+            String[] classPathAttribute = classPathAttribute(jar);
+            Path parentDir = jar.getParent();
+            for (String classPathEntry : classPathAttribute) {
+                Path otherJar = parentDir != null
+                        ? parentDir.resolve(classPathEntry)
+                        : Path.of(classPathEntry);
+                if (Files.exists(otherJar)) {
+                    // Class-Path attribute specifies that jars that
+                    // are not found are simply ignored. Do the same here
+                    classPathJars.offer(otherJar);
```

Comment:
The expansion of the class path looks right but the Class-Path attribute is awkward to deal with as its relative URI rather than a file path. More on this this later.

Reply:
Not important for now but the expansion will probably need to a "visited" set to detect cycles (yes, it is possible).

Reply:
FWIW, javac has `com.sun.tools.javac.file.FSInfo.getJarClassPath`, which is not transitive, and is implemented a bit differently, but I wonder if there's a chance for some code sharing. Maybe not now, but eventually.